falcon/asgi/request.py

```diff
@@ -794,26 +802,42 @@ def if_match(self):
         #   gets a None back on the first reference to property, it
         #   probably isn't going to access the property again (TBD).
         if self._cached_if_match is None:
-            header_value = self._asgi_headers.get('if-match')
+            header_value = self._asgi_headers.get(b'if-match')
             if header_value:
-                self._cached_if_match = helpers._parse_etags(header_value)
+                self._cached_if_match = helpers._parse_etags(header_value.decode())
 
         return self._cached_if_match
 
     @property
     def if_none_match(self):
         if self._cached_if_none_match is None:
-            header_value = self._asgi_headers.get('if-none-match')
+            header_value = self._asgi_headers.get(b'if-none-match')
             if header_value:
-                self._cached_if_none_match = helpers._parse_etags(header_value)
+                self._cached_if_none_match = helpers._parse_etags(header_value.decode())
 
         return self._cached_if_none_match
 
+    @property
+    def headers(self):
+        # NOTE(kgriffs: First time here will cache the dict so all we
+        # have to do is clone it in the future.
+        if self._cached_headers is None:
+            self._cached_headers = {
+                name.decode(): value.decode()
+                for name, value in self._asgi_headers.items()
+            }
+
+        return self._cached_headers
+
     # ------------------------------------------------------------------------
     # Public Methods
     # ------------------------------------------------------------------------
 
-    def get_header(self, name, required=False, default=None):
+    # PERF(kgriffs): Using kwarg cache, in lieu of @lru_cache on a helper method
+    #   that is then called from get_header(), was benchmarked to be more
+    #   efficient across CPython 3.6/3.8 (regardless of cythonization) and
+    #   PyPy 3.6.
```

Comment:
It would be great if someone could check this on Linux/Window to make sure this holds true across platforms.

The comparison in this case that is relevant is between `foo_a()` and `foo_c()` patterns:

```py
def foo_a(name, _cache={}):
    try:
        decoded_name = _cache[name]
    except KeyError:
        decoded_name = name.decode()
        if len(_cache) < 64:
            _cache[name] = decoded_name

    return decoded_name


@lru_cache()
def foo_b(name):
    return name.decode()


def foo_c(name):
    return foo_b(name)
```

Reply:
The difference seems too small to provide a meaningful difference (meaning that depending in the run one is better than the other:
```
In [13]: %timeit [foo_c(b'foobar%d'%i) for i in range(200)]
124 µs ± 4.44 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [14]: %timeit [foo_a(b'foobar%d'%i) for i in range(200)]
125 µs ± 7.51 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [15]: %timeit [foo_a(b'foobar%d'%i) for i in range(200)]
124 µs ± 5.32 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [17]: %timeit [foo_c(b'foobar%d'%i) for i in range(200)]
125 µs ± 3.6 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

NOTE: I did place a 64 in the lru cache

Reply:
Interesting. The kwarg pattern is definitely faster for me both on macOS and from within a pypy docker container.